cmd/snap-confine/mount-support.c

```diff
@@ -404,6 +404,49 @@ static void sc_bootstrap_mount_namespace(const struct sc_mount_config *config)
 			    dst);
 		}
 	}
+	// Since we mountec /etc from the host filesystem the /etc/os-release file
+	// (which may not be present) may be a symbolic link. It is expected that
+	// we see the real /etc/os-release file so let's do our best to show
+	// /etc/os-release as it looks like on the clasic distribution.
+	//
+	// If the symbolic link in /etc is pointing to /usr/lib/os-release then
+	// bind mount /usr/lib/os-release over itself in the core snap. This way
+	// the symlink (which we cannot over-bind mount anything) will work as
+	// expected.
+	//
+	// https://bugs.launchpad.net/canonical-livepatch-client/+bug/1667470
+	if (config->on_classic) {
+		const char *etc_os_release = "/etc/os-release";
+		const char *expected_symlink = "../usr/lib/os-release";
+		const char *usr_lib_os_release = "/usr/lib/os-release";
+		char link_target[PATH_MAX];
+		// Try to read link target from /etc/os-release and special case EINVAL
+		// instead of trying to lstat it first.
```

Comment:
This comment should indicate when readlink() will return EINVAL (eg, when not a symlink (or bufsize is negative)) and say why it is ok to special case (but before you do that, see below).